deepecho/models/basic_gan.py

```diff
@@ -0,0 +1,575 @@
+        """Obtain the category that corresponds to the highest one-hot value."""
+        max_value = float('-inf')
+        for category, idx in properties['indices'].items():
+            value = tensor[row, 0, idx]
+            if value > max_value:
+                max_value = value
+                selected = category
+
+        return selected
+
+    def _value_to_tensor(self, tensor, value, properties):
+        """Update the tensor according to the value and properties."""
+        column_type = properties['type']
+        if column_type in ('continuous', 'count'):
+            self._normalize(tensor, value, properties)
+        elif column_type in ('categorical', 'ordinal'):
+            self._one_hot_encode(tensor, value, properties)
+
+        else:
+            raise ValueError()   # Theoretically unreachable
+
+    def _data_to_tensor(self, data):
+        """Convert the input data to the corresponding tensor.
+
+        If ``self._fixed_length`` is ``False``, add a 1.0 to indicate
+        the sequence end and pad the rest of the sequence with 0.0s.
+        """
+        tensors = []
+        num_rows = len(data[0])
+        for row in range(num_rows):
+            tensor = torch.zeros(self._model_data_size)
+            for column, properties in self._data_map.items():
+                value = data[column][row]
+                self._value_to_tensor(tensor, value, properties)
+
+            tensors.append(tensor)
+
+        if not self._fixed_length:
+            tensors[-1][-1] = 1.0
+
+        for _ in range(self._max_sequence_length - num_rows):
+            tensors.append(torch.zeros(self._model_data_size))
+
+        return torch.stack(tensors, dim=0)
+
+    def _context_to_tensor(self, context):
+        """Convert the input context to the corresponding tensor."""
+        tensor = torch.zeros(self._context_size)
+        for column, properties in self._context_map.items():
+            value = context[column]
+            self._value_to_tensor(tensor, value, properties)
+
+        return tensor
+
+    def _tensor_to_data(self, tensor):
+        """Rebuild a valid sequence from the given tensor."""
+        sequence_length, num_sequences, _ = tensor.shape
+        assert num_sequences == 1
+
+        data = [None] * len(self._data_map)
+        for column, properties in self._data_map.items():
+            column_type = properties['type']
+
+            column_data = []
+            data[column] = column_data
+            for row in range(sequence_length):
+                if column_type in ('continuous', 'count'):
+                    value = self._denormalize(tensor, row, properties)
+                elif column_type in ('categorical', 'ordinal'):
+                    value = self._one_hot_decode(tensor, row, properties)
+                else:
+                    raise ValueError()   # Theoretically unreachable
+
+                if value is not None and column_type in ('count', 'ordinal'):
```

Comment:
I don't think you need to round the ordinal values?

Reply:
Thanks! You are totally right. I changed it to round only `count` values and also fixed another problem that I spotted on the sequence generation that converted ordinal integers to float values, which made the outputed categories appear as float values.